data_model/src/query/predicate/mod.rs

```diff
@@ -0,0 +1,208 @@
+#![warn(missing_docs)] // TODO
+#![allow(missing_copy_implementations)]
+//! Predicate-related logic. Should contain predicate-related `impl`s.
+
+pub mod predicate_ast_extensions;
+pub mod predicate_atoms;
+pub mod predicate_combinators;
+pub mod projectors;
+pub mod prototypes;
+
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, vec, vec::Vec};
+
+use super::*;
+
+/// Trait for generic predicates.
+pub trait PredicateTrait<T: ?Sized> {
+    /// The result of applying the predicate to a value.
+    fn applies(&self, input: &T) -> bool;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub enum CompoundPredicate<Atom> {
+    Atom(Atom),
+    Not(Box<Self>),
+    And(Vec<Self>),
+    Or(Vec<Self>),
+}
+
+impl<Atom> CompoundPredicate<Atom> {
+    /// A compound predicate that always evaluates to `true`.
+    pub const PASS: Self = Self::And(Vec::new());
+    /// A compound predicate that always evaluates to `false`.
+    pub const FAIL: Self = Self::Or(Vec::new());
+
+    pub fn and(self, other: Self) -> Self {
```

Comment:
Should `or` be added as well?

Reply:
Added along with `not` method, and also improved representation of and/or chains by more aggressively flattening them and actually using these methods in the AST combinator normalization routines.